### PR TITLE
Fix application plan changes with different services in the API

### PR DESCRIPTION
**Pull request by guicassolato**
**What this PR does / why we need it**

It prevents application plan changes between different services of the same provider through the API.

**Which issue(s) this PR fixes** 

Closes [THREESCALE-1869](https://issues.jboss.org/browse/THREESCALE-1869)

**Verification steps** 

As a provider with more than a one service containing application plans, use the 3scale API to change a buyer application to a plan that belongs to a service different than the current service the application is associated with through its current plan. For example,

- Provider `X` has services `S1` and `S2`
- `S1` issued application plan `P1`
- `S2` issued application plan `P1`
- Developer `Y` has an application `A` currently on plan `P1`
- Provider `X` performs an API PUT request to `/admin/api/accounts/<Y:id>/applications/<A:id>/change_plan.xml` with payload `{ plan_id: <P2:id> }`

The response should now be a 404, as `P2` does not belong to `S1`.

### CODE REVIEW

File: app/controllers/admin/api/buyers_applications_controller.rb [hunk 1]
```diff
@@ -158,6 +158,7 @@ def find
   ##~ op.parameters.add :name => "plan_id", :description => "ID of the new application plan.", :dataType => "int", :paramType => "query", :required => true, :threescale_name => "application_plan_ids"
   #
   def change_plan
+    application_plan = same_service_application_plans.find(params[:plan_id])
```
Comment (hallelujah) [inline app/controllers/admin/api/buyers_applications_controller.rb]: Well that will return a 404.
Probably a 412 as suggested by @mikz  because if the plan exists then we should return an error instead.
I would do it in the `#change_plan!` instead of here

Comment (guicassolato) [reply app/controllers/admin/api/buyers_applications_controller.rb]: I agree that 404 can be misleading and that anything enforced inside `#change_plan!` will be more effective.

Not sure about the 412 though. A 412 response usually means that the request does not match a precondition stated by the client itself in the request headers (in contraposition to something obscure defined by the server).

File: app/models/cinstance.rb [hunk 2]
```diff
@@ -380,6 +380,13 @@ def app_plan_change_should_request_credit_card?
     service.plan_change_permission(ApplicationPlan) == :request_credit_card
   end
 
+  class DifferentServicesPlanChangeError < StandardError; end
+
+  def change_plan!(new_plan)
```
Comment (hallelujah) [inline app/models/cinstance.rb]: Why not a simple error in validation?

File: app/models/cinstance.rb [hunk 3]
```diff
@@ -409,6 +411,10 @@ def change_plan_internal(new_plan)
     end
   end
 
+  def same_service
+    errors.add(:plan_id, :not_allowed) if plan&.service != service
```
Comment (hallelujah) [inline app/models/cinstance.rb]: AccountPlan does not have service.

Comment (guicassolato) [reply app/models/cinstance.rb]: That's only for applications.

Comment (hallelujah) [reply app/models/cinstance.rb]: I mean this will be always true for AccountPlan:
`plan&.service != service`

Comment (hallelujah) [reply app/models/cinstance.rb]: My bad, you are right, the account_contract will not have any service either.

Comment (guicassolato) [reply app/models/cinstance.rb]: Also, this is in `Cinstance`. It does not apply to all types of contracts.

File: app/controllers/admin/api/buyers_applications_controller.rb (hunk 1 above)
Comment (guicassolato) [reply app/controllers/admin/api/buyers_applications_controller.rb]: [Now](https://github.com/3scale/porta/pull/568/commits/774f9676967f18dcda839e1c187fb46e029e08cc) it turns out it will be 422  :-)

File: app/models/cinstance.rb (hunk 3 above)
Comment (Martouta) [inline app/models/cinstance.rb]: I don't understand why `plan&.service != service` instead of `plan.service != service` because this is invoked in this case: `validate :same_service, on: :update, if: :plan_id_changed?`
How can `plan_id_changed?` be `true` but at the same time `plan` being `nil` for an `update` 😕

Comment (guicassolato) [reply app/models/cinstance.rb]: The new plan can be `nil`. There's actually a test for that. And in case you're wondering, no, it should not ever be `nil`, but that's covered by other validation: https://github.com/3scale/porta/blob/3be766f22eeb2fb727449e768bf4e2487583769e/app/models/cinstance.rb#L79

Comment (Martouta) [reply app/models/cinstance.rb]: Ok, I see... before I said that, I tried your test in `test/unit/cinstance_test.rb:680` doing that change that I said and it worked so I was confused.

@guicassolato Can you tell me where is this test that you mentioned?
> There's actually a test for that

I ask because I've seen something interesting. If I run this test that I've just written right below the one you have in `test/unit/cinstance_test.rb:680` right now :
```ruby
test 'checking what happens when updating plan to nil xD' do
  pp @cinstance.change_plan! nil
  pp @cinstance.change_plan nil
  pp @cinstance.valid?
  pp @cinstance.plan
  pp @cinstance.errors.full_messages
end
```
And what I see is no errors of any kind:
image
Even if I change in this line the `plan&.service != service` for `plan.service != service` I see exactly the same 🤔 And I see why in the methods to change plan:
https://github.com/3scale/porta/blob/774f9676967f18dcda839e1c187fb46e029e08cc/app/models/contract.rb#L187-L201
https://github.com/3scale/porta/blob/774f9676967f18dcda839e1c187fb46e029e08cc/app/models/contract.rb#L263-L264
so basically `changed` is `true` when is `nil` although it didn't actually change anything so it just returns the old plan and does not have any error anywhere just like if had done the change... which doesn't look like a good idea to me  😕

Comment (guicassolato) [reply app/models/cinstance.rb]: Mine fails at https://github.com/3scale/porta/blob/3be766f22eeb2fb727449e768bf4e2487583769e/test/unit/cinstance_test.rb#L8

screen shot 2019-01-31 at 3 10 10 pm

File: test/unit/cinstance_test.rb [hunk 4]
```diff
@@ -683,6 +683,12 @@ class ChangePlanTest < ActiveSupport::TestCase
       end
       assert_includes @cinstance.errors['plan_id'], 'not allowed in this context'
     end
+
+    test 'cannot change plan to not plan at all' do
+      previous_plan = @cinstance.plan_id
+      assert_nothing_raised { @cinstance.change_plan! nil }
```
Comment (Martouta) [inline test/unit/cinstance_test.rb]: This is not what I meant 😝 This is what I meant: https://github.com/3scale/porta/pull/580
Feel free to either merge it or close it or to do what you consider with it... but it is easier to do it and explain it in the description there than adding a big abstract comment here.

Comment (mikz) [reply test/unit/cinstance_test.rb]: BTW `assert_nothing_raised` is completely useless :)
